packages/engine/src/worker/runner/python/runner.py

```diff
@@ -1,68 +1,281 @@
+        to initialize experiment-level context (`self.experiment_ctx`)
+        and each package's experiment-level data (`self.pkgs`).
+        The packages' source code is not included in the message,
+        but each package's name and type is included, and its
+        source code is looked up from a path determined by its
+        name and type.
+
+        If an error occurs during a package's custom experiment-level
+        initialization (whether a package or user error), it is sent
+        to the Rust process and experiment init is stopped early and
+        `True` is returned. (`True` is returned instead of raising
+        an exception in order to distinguish runner errors from
+        package/user errors.)
+
+        User warnings are also sent to the Rust process, but do not
+        stop experiment-level init.
+
+        This function is actually only called once per Python process,
+        soon after the process starts, because the init message is
+        always the first message that the Rust process sends to a
+        Python process.
+
+        :return: Whether a package/user error occurred
+        """
         init = self.messenger.recv_init()
-        experiment_ctx = init.shared_ctx
+        self.experiment_ctx = init.shared_ctx
         for pkg_id, config in init.pkgs.items():
             self.pkgs[pkg_id] = pkg = Package(
                 name=config.name,
                 pkg_type=config.type,
-                owned_fields=config.owned_fields
+                owned_fields=[],  # TODO: Propagate `config.owned_fields` here.
             )
 
             if pkg.start_experiment is not None:
-                pkg.start_experiment(
-                    pkg.experiment, config.payload, experiment_ctx
-                )
-
-        return experiment_ctx
+                try:
+                    result = pkg.start_experiment(
+                        pkg.experiment, config.payload, self.experiment_ctx
+                    )
+                    were_user_errors = self._handle_experiment_init_result(pkg, result)
+                    if were_user_errors:
+                        # TODO: Should we kill the runner here or let the Rust process
+                        #       terminate it after receiving user errors?
+                        return True
+
+                except Exception:
+                    # Have to catch generic exception, because package could throw anything.
+                    self._handle_pkg_error(pkg, "experiment init", sys.exc_info())
+                    return True
+
+        return False
+
+    def _handle_experiment_init_result(self, pkg, result):
+        """
+        :param pkg: The package object, with experiment-level data
+        :param result: What the package's custom experiment init returned
+        :return: Whether any user errors occurred during this
+                 package's experiment init
+        """
+        if not result:  # Package didn't return anything.
```

Comment:
Do we mean `if result is None`?

Reply:
Yes

Reply:
I see the comment, can we not just change the package impls? There aren't many of them. Or is the worry errors or something?

Reply:
Indeed, I'm worried about introducing errors. As we don't have any statical analysis, it's easy to miss some edge cases. I'd prefer doing it properly with the task mentioned at `start_experiment` and leaving it until then.